### PR TITLE
Feature/udp support

### PR DESCRIPTION
**Pull request by gillyb**
This took a long time, but I finally got to doing it. :)
Testing it on a production account I created for this, and all messages were received.

### CODE REVIEW

File: lib/logzio-nodejs.js [hunk 1]
```diff
@@ -89,9 +98,28 @@ LogzioLogger.prototype._timerSend = function() {
...
+
+    for (var i=0; i<messagesLength; i++) {
+
+        var msg = this.messages[i];
+        msg.token = this.token;
+        var buff = new Buffer(stringifySafe(msg));
+
+        this.udpClient.send(buff, 0, buff.length, this.port, this.host, function(err, bytes) {
+            if (err) 
+                throw new Error('Failed to send udp log message. err = ' + err);
```
Comment (asafm) [inline lib/logzio-nodejs.js]: Throwing an exception might the application which uses you. Let's use the same methodology below and call the callback with the error message and exception

File: lib/logzio-nodejs.js [hunk 2]
```diff
@@ -89,9 +98,28 @@ LogzioLogger.prototype._timerSend = function() {
...
+        });
+    }
```
Comment (asafm) [inline lib/logzio-nodejs.js:123]: Maybe add debug - "Sent x messages via UDP" ?

File: lib/logzio-nodejs.js [hunk 3]
```diff
@@ -1,20 +1,25 @@
...
+
+    if (options.protocol === 'udp') {
+        this.udpClient = dgram.createSocket('udp4');
+    }
 
     this.token = options.token;
     this.protocol = options.protocol || 'http';
     this.host = options.host || 'listener.logz.io';
-    this.port = this.protocol == 'http' ? 8070 : 8071;
+    this.port = this.protocol === 'udp' ? 5050 : (this.protocol === 'http' ? 8070 : 8071);
```
Comment (asafm) [inline lib/logzio-nodejs.js]: I think it's nicer to have a map of protocols to ports. 
We can re-use this map to validate allowed protocols as well above

File: README.md [hunk 4]
```diff
@@ -31,7 +31,7 @@ logger.log(obj);
 * **token** 
     Mandatory. Your API logging token. Look it up in the Device Config tab in Logz.io
 * **type** - Log type. Help classify logs into different classifications
-* **protocol** - 'http' or 'https'. Default: http
+* **protocol** - 'http', 'https' or 'udp'. Default: http
```
Comment (asafm) [inline README.md:34]: I think when we introduce udp, there are different sets of params. For example: bufferSize and numberOfRetries isn't not relevant. Maybe we should make it clear in the documentation which params are for http and which for UDP.

File: lib/logzio-nodejs.js [hunk 5]
```diff
@@ -128,9 +156,17 @@ LogzioLogger.prototype.log = function(msg) {
 };
 
 LogzioLogger.prototype._popMsgsAndSend = function() {
-    var bulk = this._createBulk(this.messages);
-    this._debug('Sending bulk #' + bulk.id);
-    this._send(bulk);
+    
+    if (this.protocol === 'udp') {
```
Comment (asafm) [inline lib/logzio-nodejs.js:169]: Since UDP is not bulk, I would send the UDP message immediately as it logged.

File: README.md [hunk 6]
```diff
@@ -41,6 +41,9 @@ logger.log(obj);
 * **addTimestampWithNanoSecs** - Add a timestamp with nano seconds granularity. This is needed when many logs are sent in the same millisecond, so you can properly order the logs in kibana. The added timestamp field will be `@timestamp_nano` Default: false
 
 ## Update log
+**0.3.11**
```
Comment (asafm) [inline README.md]: I think UDP calls for 0.4

**Review by asafm — CHANGES_REQUESTED** [review]
Looks good - left some comments.
The prominent thing missing here is testing. We can do the same we did for http and use a mock udp server

File: lib/logzio-nodejs.js (hunk 5 above)
Comment (gillyb) [reply lib/logzio-nodejs.js:169]: I thought about that, but then thought maybe some people will still want to send all the udp messages once every 60 seconds, so their network isn't always sending udp.
If they still want it to send each message they can use `bufferSize=1` - I will add this to the documentation.

File: README.md (hunk 6 above)
Comment (gillyb) [reply README.md]: 👍

**Comment by gillyb** [comment]
@asafm I went over the comments, and implemented them all. I also added two unit tests to test the new UDP functionality. In addition, I added some more info to the documentation regarding usage of udp, and bumped the version to 0.4.
You want to take a look, can I merge this ?

**Review by asafm — COMMENTED** [review]
Looks good.
Regarding the test it self - I prefer to avoid mocking since it exposes internal code. Maybe a simple udp server would do the trick - I saw npm has one for this very much like we use in the http testing.

**Pull request merged**